### PR TITLE
feat: Add support for RLike

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

Regular expression support is usually important in ETL jobs, so we should start adding support.

## What changes are included in this PR?

Add support for RLike.

## How are these changes tested?

New test.

### CODE REVIEW

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 1]
```diff
@@ -1094,24 +1094,46 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
             None
           }
 
-        // TODO waiting for arrow-rs update
-//      case RLike(left, right) =>
```
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1098]: I forgot the details. I remember this was added by someone at the early development. Not sure if we have anything update in arrow-rs for it.

Comment (kazuyukitanimura) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1098]: cc @snmvaughan

**Comment by andygrove** [comment]
Could I get more feedback on this PR @parthchandra @kazuyukitanimura @huaxingao

File: core/src/execution/datafusion/expressions/regexp.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+impl PartialEq<dyn Any> for RLike {
+    fn eq(&self, other: &dyn Any) -> bool {
+        down_cast_any_ref(other)
+            .downcast_ref::<Self>()
+            .map(|x| self.child.eq(&x.child) && self.pattern.eq(&x.pattern))
+            .unwrap_or(false)
+    }
+}
+
+impl PhysicalExpr for RLike {
```
Comment (kazuyukitanimura) [inline core/src/execution/datafusion/expressions/regexp.rs:65]: In `strings.rs` we have
```
// make_predicate_function!(RLike, rlike_dyn, rlike_utf8_scalar_dyn);
```
Is it possible to use it instead?

Comment (andygrove) [reply core/src/execution/datafusion/expressions/regexp.rs:65]: Arrow does not provide an implementatin of `rlike_dyn` or `rlike_utf8_scalar_dyn`

Comment (andygrove) [reply core/src/execution/datafusion/expressions/regexp.rs:65]: Also, I am expecting us to start specializing our implementation to make it compatible with Spark, so it seems better to have our own implementation.

File: core/src/execution/datafusion/expressions/regexp.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            } else {
+                Err(
+                    CometError::Internal("Only scalar regex patterns are supported".to_string())
+                        .into(),
+                )
+            }
+        } else {
+            // this should be unreachable because Spark will evaluate regex expressions against
+            // literal strings as part of query planning
+            Err(CometError::Internal("Only columnar inputs are supported".to_string()).into())
```
Comment (kazuyukitanimura) [inline core/src/execution/datafusion/expressions/regexp.rs:120]: Is it possible to encounter dictionary type?

File: core/src/execution/datafusion/expressions/regexp.rs (hunk 2 above)
Comment (kazuyukitanimura) [reply core/src/execution/datafusion/expressions/regexp.rs:65]: Got it. For other functions e.g. `like`, we contributed in `arrow-rs` https://github.com/apache/arrow-rs/pull/3502 in the past...

Comment (andygrove) [reply core/src/execution/datafusion/expressions/regexp.rs:65]: I think it makes sense for Like. Regular expression support is just going to be quite challenging if we choose to try and mimic the behavior of the java regexp engine.

File: core/src/execution/datafusion/expressions/regexp.rs (hunk 3 above)
Comment (andygrove) [reply core/src/execution/datafusion/expressions/regexp.rs:120]: Good point. Yes, it probably is.

**Pull request closed without merging**